### PR TITLE
FIX-#4057: Allow reading an empty parquet file.

### LINKED ISSUE #4057: Can't deserialize an empty Modin DataFrame using `read_parquet`
### System information
Modin 0.12.0
Ray 1.7.1
python 3.9.7
pyarrow 6.0.1

### Describe the problem
```
df = modin_pd.DataFrame([])
out = io.BytesIO()
df.to_parquet(path=out, engine="pyarrow", compression="snappy")

out.seek(0)
print(out.read())

pd.read_parquet(path=out, engine="pyarrow")
```

throws

```
b'PAR1\x15\x04\x15\x00\x15\x02L\x15\x00\x15\x04\x12\x00\x00\x00&&\x1c\x15\x02\x195\x04\x00\x06\x19\x18\x11__index_level_0__\x15\x02\x16\x00\x16\x1c\x16\x1e&\x00&\x08)\x1c\x15\x04\x15\x04\x15\x02\x00\x00\x00\x15\x02\x19,5\x00\x18\x06schema\x15\x02\x00\x15\x02%\x02\x18\x11__index_level_0__l\xbc\x00\x00\x00\x16\x00\x19\x1c\x19\x1c&&\x1c\x15\x02\x195\x04\x00\x06\x19\x18\x11__index_level_0__\x15\x02\x16\x00\x16\x1c\x16\x1e&\x00&\x08)\x1c\x15\x04\x15\x04\x15\x02\x00\x00\x00\x16\x1c\x16\x00&\x08\x16\x1e\x14\x00\x00\x19,\x18\x06pandas\x18\xf6\x02{"index_columns": ["__index_level_0__"], "column_indexes": [{"name": null, "field_name": null, "pandas_type": "empty", "numpy_type": "object", "metadata": null}], "columns": [{"name": null, "field_name": "__index_level_0__", "pandas_type": "empty", "numpy_type": "object", "metadata": null}], "creator": {"library": "pyarrow", "version": "6.0.1"}, "pandas_version": "1.3.4"}\x00\x18\x0cARROW:schema\x18\xec\x05/////ygCAAAQAAAAAAAKAA4ABgAFAAgACgAAAAABBAAQAAAAAAAKAAwAAAAEAAgACgAAAKwBAAAEAAAAAQAAAAwAAAAIAAwABAAIAAgAAACEAQAABAAAAHYBAAB7ImluZGV4X2NvbHVtbnMiOiBbIl9faW5kZXhfbGV2ZWxfMF9fIl0sICJjb2x1bW5faW5kZXhlcyI6IFt7Im5hbWUiOiBudWxsLCAiZmllbGRfbmFtZSI6IG51bGwsICJwYW5kYXNfdHlwZSI6ICJlbXB0eSIsICJudW1weV90eXBlIjogIm9iamVjdCIsICJtZXRhZGF0YSI6IG51bGx9XSwgImNvbHVtbnMiOiBbeyJuYW1lIjogbnVsbCwgImZpZWxkX25hbWUiOiAiX19pbmRleF9sZXZlbF8wX18iLCAicGFuZGFzX3R5cGUiOiAiZW1wdHkiLCAibnVtcHlfdHlwZSI6ICJvYmplY3QiLCAibWV0YWRhdGEiOiBudWxsfV0sICJjcmVhdG9yIjogeyJsaWJyYXJ5IjogInB5YXJyb3ciLCAidmVyc2lvbiI6ICI2LjAuMSJ9LCAicGFuZGFzX3ZlcnNpb24iOiAiMS4zLjQifQAABgAAAHBhbmRhcwAAAQAAABQAAAAQABQACAAGAAcADAAAABAAEAAAAAAAAQEQAAAAKAAAAAQAAAAAAAAAEQAAAF9faW5kZXhfbGV2ZWxfMF9fAAAABAAEAAQAAAA=\x00\x18\x1fparquet-cpp-arrow version 6.0.1\x19\x1c\x1c\x00\x00\x00\x1c\x05\x00\x00PAR1'
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/var/folders/9l/q_y0w5yx2_1_h65cprqgczyc0000gq/T/ipykernel_62102/810778694.py in <module>
      6 print(out.read())
      7 
----> 8 pd.read_parquet(path=out, engine="pyarrow")

lib/python3.9/site-packages/modin/pandas/io.py in read_parquet(path, engine, columns, storage_options, use_nullable_dtypes, **kwargs)
    214     Engine.subscribe(_update_engine)
    215     return DataFrame(
--> 216         query_compiler=FactoryDispatcher.read_parquet(
    217             path=path,
    218             engine=engine,

b/python3.9/site-packages/modin/core/execution/dispatching/factories/dispatcher.py in read_parquet(cls, **kwargs)
    173     @_inherit_docstrings(factories.BaseFactory._read_parquet)
    174     def read_parquet(cls, **kwargs):
--> 175         return cls.__factory._read_parquet(**kwargs)
    176 
    177     @classmethod

python3.9/site-packages/modin/core/execution/dispatching/factories/factories.py in _read_parquet(cls, **kwargs)
    195     )
    196     def _read_parquet(cls, **kwargs):
--> 197         return cls.io_cls.read_parquet(**kwargs)
    198 
    199     @classmethod

python3.9/site-packages/modin/core/io/file_dispatcher.py in read(cls, *args, **kwargs)
    149         postprocessing work on the resulting query_compiler object.
    150         """
--> 151         query_compiler = cls._read(*args, **kwargs)
    152         # TODO (devin-petersohn): Make this section more general for non-pandas kernel
    153         # implementations.

python3.9/site-packages/modin/core/io/column_stores/parquet_dispatcher.py in _read(cls, path, engine, columns, **kwargs)
    100                 column_names = [c for c in column_names if c not in index_columns]
    101             columns = [name for name in column_names if not PQ_INDEX_REGEX.match(name)]
--> 102         return cls.build_query_compiler(path, columns, **kwargs)

python3.9/site-packages/modin/core/io/column_stores/column_store_dispatcher.py in build_query_compiler(cls, path, columns, **kwargs)
    214             Query compiler with imported data for further processing.
    215         """
--> 216         col_partitions, column_widths = cls.build_columns(columns)
    217         partition_ids = cls.call_deploy(path, col_partitions, **kwargs)
    218         index, row_lens = cls.build_index(partition_ids)

python3.9/site-packages/modin/core/io/column_stores/column_store_dispatcher.py in build_columns(cls, columns)
    169         col_partitions = [
    170             columns[i : i + column_splits]
--> 171             for i in range(0, len(columns), column_splits)
    172         ]
    173         column_widths = [len(c) for c in col_partitions]

ValueError: range() arg 3 must not be zero
```

### PR DESCRIPTION
**Pull request by mvashishtha**
Signed-off-by: mvashishtha <mahesh@ponder.io>

## What do these changes do?

Allow reading an empty parquet file.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/developer/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4057
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 1]
```diff
@@ -164,9 +166,11 @@ def build_columns(cls, columns):
             if len(columns) % num_partitions == 0
             else len(columns) // num_partitions + 1
         )
+        # If columns is empty, column_splits == 0, but the third argument, the
+        # step, in the range constructor must be nonzero, so use 1 instead.
         col_partitions = [
             columns[i : i + column_splits]
-            for i in range(0, len(columns), column_splits)
+            for i in range(0, len(columns), max(column_splits, 1))
```
Comment (RehanSD) [inline modin/core/io/column_stores/column_store_dispatcher.py:173]: If we are reading an empty parquet file, `len(columns)` and `column_splits` should both be 0, right? Maybe in that case, it makes sense to have a short circuiting if statement that just creates an empty list like so?
```python
if len(columns) == 0:
    col_partitions = [[]]
else:
...
```

**Review by RehanSD — CHANGES_REQUESTED** [review]
Left some comments - I'm wondering if its possible to short circuit when reading the metadata from the parquet file and notice its empty and avoid hitting the build_index codepath altogether?

File: modin/core/io/column_stores/column_store_dispatcher.py (hunk 1 above)
Comment (mvashishtha) [reply modin/core/io/column_stores/column_store_dispatcher.py:173]: Done.

**Comment by mvashishtha** [comment]
> I'm wondering if its possible to short circuit when reading the metadata from the parquet file and notice its empty and avoid hitting the build_index codepath altogether?

@RehanSD That's possible, but I think to do that we'd have to change each column store dispatcher at a line like [this](https://github.com/modin-project/modin/blob/5ca6d40f0ab9e6fede0dd5e77a415231a394fa90/modin/core/io/column_stores/parquet_dispatcher.py#L80) to always read the metadata and then create an empty query compiler if there are no columns that can be read. I think it's simpler to let the base class handle the case where there are no columns to read, as I've done here.

File: docs/release_notes/release_notes-0.14.0.rst [hunk 2]
```diff
@@ -4,7 +4,7 @@ Key Features and Updates
 ------------------------
 
 * Stability and Bugfixes
-  *
+  * Allow reading an empty parquet file (5767128)
```
Comment (RehanSD) [inline docs/release_notes/release_notes-0.14.0.rst:7]:
```suggestion
  * FIX-#4057: Allow reading an empty parquet file (#4075)
```

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 3]
```diff
@@ -122,7 +122,9 @@ def build_index(cls, partition_ids):
             List with lengths of index chunks.
         """
         num_partitions = NPartitions.get()
-        index_len = cls.materialize(partition_ids[-2][0])
+        index_len = (
```
Comment (RehanSD) [inline modin/core/io/column_stores/column_store_dispatcher.py:125]: Is it worth shortcutting here? i.e.:
```python
if len(partition_ids) == 0:
      return pd.RangeIndex(0), 0
....
```
?

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 4]
```diff
@@ -158,16 +160,20 @@ def build_columns(cls, columns):
...
+        else:
+            num_partitions = NPartitions.get()
+            column_splits = (
+                columns_length // num_partitions
+                if columns_length % num_partitions == 0
+                else columns_length // num_partitions + 1
+            )
+            col_partitions = [
+                columns[i : i + column_splits]
+                for i in range(0, len(columns), column_splits)
```
Comment (RehanSD) [inline modin/core/io/column_stores/column_store_dispatcher.py:175]:
```suggestion
                for i in range(0, columns_length, column_splits)
```

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 5]
```diff
@@ -158,16 +160,20 @@ def build_columns(cls, columns):
...
-            if len(columns) % num_partitions == 0
-            else len(columns) // num_partitions + 1
-        )
-        col_partitions = [
-            columns[i : i + column_splits]
-            for i in range(0, len(columns), column_splits)
-        ]
+        columns_length = len(columns)
+        if columns_length == 0:
+            col_partitions = []
```
Comment (RehanSD) [inline modin/core/io/column_stores/column_store_dispatcher.py:165]:
```suggestion
            return [], []
```

Comment (RehanSD) [reply modin/core/io/column_stores/column_store_dispatcher.py:165]: And remove `else` below?

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 6]
```diff
@@ -158,16 +160,20 @@ def build_columns(cls, columns):
...
+        else:
+            num_partitions = NPartitions.get()
+            column_splits = (
+                columns_length // num_partitions
+                if columns_length % num_partitions == 0
+                else columns_length // num_partitions + 1
+            )
+            col_partitions = [
```
Comment (RehanSD) [inline modin/core/io/column_stores/column_store_dispatcher.py:173]: This is potentially more trouble than its worth, but we could do this:
```python
col_partitions, column_widths = zip(*[
                 (columns[i : i + column_splits], len(columns[i : i + column_splits]))
                 for i in range(0, len(columns), column_splits)
])
return list(col_partitions), list(column_widths])
```
but this makes the code slightly less readable to save an additional map over the columns which idk if its worth it.

**Review by RehanSD — CHANGES_REQUESTED** [review]
Left some comments, but looks good to me!

File: modin/core/io/column_stores/column_store_dispatcher.py (hunk 6 above)
Comment (mvashishtha) [reply modin/core/io/column_stores/column_store_dispatcher.py:173]: I wouldn't expect that to improve performance much. Both versions do the same map to get the column splits, and the step of computing lengths as we do now should be fairly fast as it's just a map over a list of size `NPartitions`. `zip` has to go through all of the new splits, too.

Anyway, this is unrelated to this PR, but let me know offline if you want to talk about it more.

File: modin/pandas/test/test_io.py [hunk 7]
```diff
@@ -1387,6 +1387,21 @@ def test_read_parquet_without_metadata(self):
         finally:
             teardown_test_files([parquet_fname, csv_fname])
 
+    @pytest.mark.xfail(
+        condition="config.getoption('--simulate-cloud').lower() != 'off'",
```
Comment (RehanSD) [inline modin/pandas/test/test_io.py:1391]: Does the CI fail if we don't include this xfail?

**Review by RehanSD — COMMENTED** [review]
Left a quick q about tests, but LGTM!

**Review by naren-ponder — DISMISSED** [review]

File: modin/core/io/column_stores/column_store_dispatcher.py [hunk 8]
```diff
@@ -158,11 +160,14 @@ def build_columns(cls, columns):
...
         column_splits = (
-            len(columns) // num_partitions
-            if len(columns) % num_partitions == 0
-            else len(columns) // num_partitions + 1
+            columns_length // num_partitions
+            if columns_length % num_partitions == 0
+            else columns_length // num_partitions + 1
         )
         col_partitions = [
             columns[i : i + column_splits]
```
Comment (RehanSD) [inline modin/core/io/column_stores/column_store_dispatcher.py:173]: Can the line below be changed to `columns_length` as well?

File: modin/pandas/test/test_io.py [hunk 9]
```diff
@@ -1396,6 +1396,17 @@ def test_read_parquet_without_metadata(self):
...
+    def test_read_empty_parquet_file(self):
+        test_df = pandas.DataFrame()
+
+        with tempfile.TemporaryDirectory() as directory:
+            path = f"{directory}/data"
+            os.makedirs(path)
+            test_df.to_parquet(path + "/part-00000.parquet")
+            read_df = pd.read_parquet(path)
+
+            df_equals(test_df, read_df)
```
Comment (RehanSD) [inline modin/pandas/test/test_io.py:1408]: Should we check that its equal with `pandas.read_parquet` as well?

**Review by RehanSD — CHANGES_REQUESTED** [review]
Left two minor comments, looks good to me!

File: modin/core/io/column_stores/column_store_dispatcher.py (hunk 8 above)
Comment (mvashishtha) [reply modin/core/io/column_stores/column_store_dispatcher.py:173]: Done.

File: modin/pandas/test/test_io.py (hunk 9 above)
Comment (mvashishtha) [reply modin/pandas/test/test_io.py]: Now that you mention it, I think that's the only thing we should check. I switched the equality check to an `eval_io` call which will do that.

**Review by jeffreykennethli — DISMISSED** [review]

**Review by naren-ponder — APPROVED** [review]

**Comment by devin-petersohn** [comment]
@mvashishtha Please clear conflicts

**Comment by mvashishtha** [comment]
> @mvashishtha Please clear conflicts

Done!

File: modin/pandas/test/test_io.py (hunk 7 above)
Comment (mvashishtha) [reply modin/pandas/test/test_io.py]: Nope, I deleted it.

**Review by RehanSD — APPROVED** [review]
LGTM! Great work!

**Pull request merged**